// openfsm-address-dto/AddressDto.js
class AddressDto {
    constructor(data = {}) {
        this._address_id = data.addressId || data.address_id || undefined;
        this._fias_id = data.fiasId || data.fias_id || undefined;
        this._fias_level = data.fiasLevel || data.fias_level || undefined;
        this._value = data.value || undefined;
        this._city = data.city ||  undefined;
        this._country = data.country || undefined;
        this._flat = data.flat ||  undefined;
        this._house = data.house || undefined;
        this._postal_code = data.postalCode || data.postal_code || undefined;
        this._region = data.region ||  undefined;
        this._street = data.street ||  undefined;
        this._created_at = data.createdAt || data.created_at || undefined;
        this._deleted_at = data.deletedAt || data.deleted_at || undefined;
        this._is_default = data.isDefault || data.is_default || undefined;
        this._latitude = data.latitude || data.latitude || undefined;
        this._longitude = data.longitude || data.longitude || undefined;
    }

    // Геттеры
    getIsDefault() {
        return this._is_default;
    }

    getLatitude() {
        return this._latitude;
    }
    getLongitude() {
        return this._longitude;
    }

    getAddressId() {
        return this._address_id;
    }

    getFiasId() {
        return this._fias_id;
    }

    getFiasLevel() {
        return this._fias_level;
    }

    getValue() {
        return this._value;
    }

    getCity() {
        return this._city;
    }

    getCountry() {
        return this._country;
    }

    getFlat() {
        return this._flat;
    }

    getHouse() {
        return this._house;
    }

    getPostalCode() {
        return this._postal_code;
    }

    getRegion() {
        return this._region;
    }

    getStreet() {
        return this._street;
    }

    getCreatedAt() {
        return this._created_at;
    }

    getDeletedAt() {
        return this._deleted_at;
    }

    // Сеттеры
    setIsDefault(value) {
        this._is_default  = value;
    }

    setLatitude(value) {
        this._latitude = value;
    }

    setLongitude(value) {
        this._longitude = value;
    }

    setAddressId(value) {
        this._address_id = value;
    }

    setFiasId(value) {
        this._fias_id = value;
    }

    setFiasLevel(value) {
        this._fias_level = value;
    }

    setValue(value) {
        this._value = value;
    }

    setCity(value) {
        this._city = value;
    }

    setCountry(value) {
        this._country = value;
    }

    setFlat(value) {
        this._flat = value;
    }

    setHouse(value) {
        this._house = value;
    }

    setPostalCode(value) {
        this._postal_code = value;
    }

    setRegion(value) {
        this._region = value;
    }

    setStreet(value) {
        this._street = value;
    }

    setCreatedAt(value) {
        this._created_at = value;
    }

    setDeletedAt(value) {
        this._deleted_at = value;
    }

    // Метод для преобразования объекта в JSON
    toJSON() {
        return {
            addressId: this._address_id,
            fiasId: this._fias_id,
            fiasLevel: this._fias_level,
            value: this._value,
            city: this._city,
            country: this._country,
            flat: this._flat,
            house: this._house,
            postalCode: this._postal_code,
            latitude: this._latitude,
            longitude: this._longitude,
            region: this._region,
            street: this._street,
            createdAt: this._created_at,
            deletedAt: this._deleted_at,
            isDefault: this._is_default,
        };
    }
}

module.exports = AddressDto;
