// openfsm-address-dto/AddressDto.js
class AddressDto {
    constructor(data = {}) {
        this._address_id = data.addressId || undefined;
        this._fias_id = data.fiasId || undefined;
        this._fias_level = data.fiasLevel || undefined;
        this._value = data.value || undefined;
        this._city = data.city || undefined;
        this._country = data.country || undefined;
        this._flat = data.flat || undefined;
        this._house = data.house || undefined;
        this._postal_code = data.postalCode || undefined;
        this._region = data.region || undefined;
        this._street = data.street || undefined;
        this._created_at = data.createdAt || undefined;
        this._deleted_at = data.deletedAt || undefined;
    }

    // Геттеры
    getAddressId() {
        return this._address_id;
    }

    getFiasId() {
        return this._fias_id;
    }

    getFiasLevel() {
        return this._fias_level;
    }

    getValue() {
        return this._value;
    }

    getCity() {
        return this._city;
    }

    getCountry() {
        return this._country;
    }

    getFlat() {
        return this._flat;
    }

    getHouse() {
        return this._house;
    }

    getPostalCode() {
        return this._postal_code;
    }

    getRegion() {
        return this._region;
    }

    getStreet() {
        return this._street;
    }

    getCreatedAt() {
        return this._created_at;
    }

    getDeletedAt() {
        return this._deleted_at;
    }

    // Сеттеры
    setAddressId(value) {
        this._address_id = value;
    }

    setFiasId(value) {
        this._fias_id = value;
    }

    setFiasLevel(value) {
        this._fias_level = value;
    }

    setValue(value) {
        this._value = value;
    }

    setCity(value) {
        this._city = value;
    }

    setCountry(value) {
        this._country = value;
    }

    setFlat(value) {
        this._flat = value;
    }

    setHouse(value) {
        this._house = value;
    }

    setPostalCode(value) {
        this._postal_code = value;
    }

    setRegion(value) {
        this._region = value;
    }

    setStreet(value) {
        this._street = value;
    }

    setCreatedAt(value) {
        this._created_at = value;
    }

    setDeletedAt(value) {
        this._deleted_at = value;
    }

    // Метод для преобразования объекта в JSON
    toJSON() {
        return {
            address_id: this._address_id,
            fias_id: this._fias_id,
            fias_level: this._fias_level,
            value: this._value,
            city: this._city,
            country: this._country,
            flat: this._flat,
            house: this._house,
            postal_code: this._postal_code,
            region: this._region,
            street: this._street,
            created_at: this._created_at,
            deleted_at: this._deleted_at,
        };
    }
}

module.exports = AddressDto;
